torch/jit/annotations.py

```diff
@@ -94,8 +98,13 @@ def is_vararg(the_callable):
 
 
 def get_param_names(fn, n_args):
-    if is_script_callable(fn):
+    if hasattr(fn, '__call__') and is_function_or_method(fn.__call__):
```

Comment:
even functions have `__call__` defined, and when you access `__call__` on functions the following fields are dropped:

`{'__closure__', '__code__', '__globals__', '__module__', '__dict__', '__kwdefaults__', '__get__', '__annotations__', '__defaults__'}`
 
which may mess things up later. so this current logic is probably not sufficient.